internal/expr/complete.go

```diff
@@ -30,8 +30,21 @@ func (pe *PreparedExpr) Complete(args ...any) (ce *CompletedExpr, err error) {
 			return nil, fmt.Errorf("nil parameter")
 		}
 		v := reflect.ValueOf(arg)
-		tv[v.Type()] = v
-		typeNames = append(typeNames, v.Type().String())
+		t := reflect.TypeOf(arg)
+
+		switch t.Kind() {
+		case reflect.Struct:
+			tv[t] = v
+		case reflect.Map:
+			if err := CheckValidMapType(t); err != nil {
+				return nil, err
+			}
+			tv[reflect.TypeOf(M{})] = reflect.ValueOf(arg)
```

Comment:
Good point! I've changed to use `v` instead of `reflect.ValueOf(arg)`.  

For the key, the reason why it's `reflect.TypeOf(M{})` is that `args` come from outside the `expr` package, so the type is something other than `expr.M`. However, later on in `Complete()`, we need to check the types we receive in the input `locs` and see if a type we have in an input `loc` matches any entry in `tv`, by simply doing `v, ok := tv[in.typ]`, as this is required for the argument to `Complete()` to be a valid one. Just another peculiarity we have due to not being able to do circular imports :P To make sure, I tried `tv[t] = v` for the Map, and this produces an error sadly.

For the `reflect.Struct` case, I thought it might be nice to be explicit to list the case for `Struct` in which we directly add the value as `tv[t] = v`, and as I couldn't do `tv[t] = v` for the Map case. The default case is just to make sure that anything other than a Struct or an M type is a reason to return an error. Do you reckon I should remove the `default` case?